lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,378 @@
+        {cluster_centers, indices, c}
+      else
+        {Nx.tensor(-1, type: Nx.type(data)), Nx.broadcast(Nx.tensor(-1, type: :s64), {n}),
+         Nx.broadcast(Nx.tensor(-1, type: :s64), {n})}
+      end
+
+    %__MODULE__{
+      affinity_matrix: affinity_matrix,
+      cluster_centers_indices: cluster_centers_indices,
+      cluster_centers: cluster_centers,
+      labels: labels,
+      num_clusters: k,
+      similarity_matrix: s
+    }
+  end
+
+  @doc """
+  Computes labels and centers of clusterization.
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:affinity_matrix` - Affinity matrix. It is a negated squared euclidean distance of each pair of points.
+
+    * `:clusters_centers` - Cluster centers from the initial data.
+
+    * `:cluster_centers_indices` - Indices of cluster centers.
+
+    * `:labels` - Labels of each point.
+
+    * `:num_clusters` - Number of clusters.
+
+    * `:similarity_matrix` - Similarity matrix. It is a similarity of each pair of points.
+
+  ## Examples
+
+      iex> seed = 42
+      iex> x = Nx.tensor([[12,5,78,2], [1,-5,7,32], [-1,3,6,1], [1,-2,5,2]])
+      iex> model = Scholar.Cluster.AffinityPropagation.fit(x, seed: seed)
+      iex> Scholar.Cluster.AffinityPropagation.prune(model)
+      %Scholar.Cluster.AffinityPropagation{
+        labels: Nx.tensor([0, 1, 1, 1]),
+        cluster_centers_indices: Nx.tensor([0, 3]),
+        affinity_matrix: Nx.tensor(
+          [
+            [0.0, -6162.0, -5358.0, -5499.0],
+            [-6162.0, 0.0, -1030.0, -913.0],
+            [-5358.0, -1030.0, 0.0, -31.0],
+            [-5499.0, -913.0, -31.0, 0.0]
+          ]),
+        cluster_centers: Nx.tensor(
+          [
+            [12, 5, 78, 2],
+            [1, -2, 5, 2]
+          ]
+        ),
+        num_clusters: Nx.tensor(2, type: :u64),
+        similarity_matrix: Nx.tensor(
+          [
+            [-971.5, -6162.0, -5358.0, -5499.0],
+            [-6162.0, -971.5, -1030.0, -913.0],
+            [-5358.0, -1030.0, -971.5, -31.0],
+            [-5499.0, -913.0, -31.0, -971.5]
+          ]
+        )
+      }
+  """
+  def prune(
+        %__MODULE__{
+          cluster_centers_indices: cluster_centers_indices,
+          cluster_centers: cluster_centers,
+          num_clusters: k,
+          similarity_matrix: similarity_matrix
+        } = model
+      ) do
+    k = Nx.to_number(k)
+
+    cluster_centers_indices =
+      Nx.to_flat_list(cluster_centers_indices) |> Enum.filter(&(&1 != -1)) |> Nx.tensor()
+
+    cluster_centers = Nx.take(cluster_centers, cluster_centers_indices)
+
+    labels = Nx.argmax(Nx.take(similarity_matrix, cluster_centers_indices, axis: 1), axis: 1)
+    labels = Nx.indexed_put(labels, Nx.new_axis(cluster_centers_indices, -1), Nx.iota({k}))
+
+    %__MODULE__{
+      model
+      | cluster_centers_indices: cluster_centers_indices,
+        cluster_centers: cluster_centers,
+        labels: labels
+    }
```

Comment:
So here is what I am thinking: could we make it so the labels get `-1` and the cluster centers get NaNs in the positions where indices are `-1`? This way when we compute the euclidian distance and then discard NaNs?

Reply:
Labels are the labels of data, so this wouldn't make sense. For cluster centers I originally pick nans, but eventually resign, because nans are available only as floats. If we fill them with 0 we can preserve the correct type of data.

Reply:
I mean at this point we can put either all -1 as labels or try to calculate them correctly.

Reply:
The issue with filling them with `0` is that then the eucledian distance can accidentally match on them, no? Or would we do a pass to discard them later?

Reply:
After the `prune` function all parameters are ok, so our problem is to compute labels without knowing explicitly how many clusters do we have. 

For example, suppose we have given from the algorithm that the clusters are samples on indices [2, 6, 7, 11] and num_samples = 15. Ideally, if the data to fit has the following closest centroids [2, 2, 6, 7, 11, 2 ...] we want to label them as [0, 0, 1, 2, 3, 0 ... ], but we couldn't achieve this because of static shapes. If we could somehow return labels [2, 2, 6, 7, 11, 2 ...] this will also be great and we relabel the data in `prune` to  [0, 0, 1, 2, 3, 0 ... ]. Currently, in `fit` I tried to achieve this, but some labels are mismatched.

Reply:
> The issue with filling them with `0` is that then the eucledian distance can accidentally match on them, no? Or would we do a pass to discard them later?

Since we got the whole information in `cluster_centers_indices`, we never make a mistake in `prune` function. However, 0s are confusing because someone may think that those 0s are legit clusters with position [0, 0, ..., 0]

Reply:
Maybe we force affinity propagation to convert to floats for now?

Reply:
Yeah, then we could use Nans, I think that's a good idea.

Reply:
How about adding a fake cluster at the end of the cluster list?
Then pruning is just a matter of dropping that last position from the shape and all related data.